### PR TITLE
core(start-url): stay offline for entirety of offlinePass

### LINKED ISSUE #9450: Offline pass doesn't always go offline
#### Provide the steps to reproduce
1. Run LH on https://www.bhphotovideo.com/
 

#### What is the current behavior?

Lighthouse reports that the start URL does not load offline but the current page does.

You can observe that in the offline pass, the desktop version of the page is loaded, and upon inspecting lighthouse internals see that the `fromServiceWorker` and `fromDiskCache` flags on all network requests are `false`. This shouldn't be possible when offline.

#### What is the expected behavior?

Lighthouse reports that the start URL loads offline. A page like the below should be visible during the offline pass, not the desktop version of the site.

image

#### Environment Information
* Affected Channels: CLI
* Lighthouse version: 5.2.0
* Node.js version: v10.15.3
* Operating System: macOS

**Related issues**
brought over from https://github.com/GoogleChrome/lighthouse/issues/2688#issuecomment-514738794

### PR DESCRIPTION
**Pull request by patrickhulce**
**Summary**
tl;dr - `StartUrl` gatherer is not actually offline while it is checking for service worker start URL fetching. This PR fixes that.

In #6957, the start url gatherer was refactored to allow WebAppManifest to be part of core artifacts and avoid being double fetched. The code appeared to assume that the page was offline at the start of the `afterPass` because [it went back online](https://github.com/GoogleChrome/lighthouse/commit/fb6c376feb8ee7161137baf5f86e978b4f9d7bfc#diff-dd5e1188fbf8a1aa7763695960187be1L21) to fetch the manifest before going back offline. This wasn't actually the case though. Due to refactorings long ago in gather runner we actually always run the entirety of `afterPass` while online.

For pages with service workers that are always controlling the page this doesn't end up making much of a difference because of good ol' #709. However, if the service worker only handles requests while the page is in an offline state (not really clear *how* one does this), then all requests actually go out over the network and we trip our `!response.fromServiceWorker` line to falsely fail the page (which is normally a good thing because it means that all pages that didn't have a service worker didn't magically start passing when this bug was introduced! 😄)

**Related Issues/PRs**
fixes #9450

### CODE REVIEW

File: lighthouse-core/gather/gatherers/offline.js [hunk 1]
```diff
@@ -21,14 +21,13 @@ class Offline extends Gatherer {
    * @param {LH.Gatherer.LoadData} loadData
    * @return {Promise<LH.Artifacts['Offline']>}
    */
-  afterPass(passContext, loadData) {
+  async afterPass(passContext, loadData) {
```
Comment (brendankenny) [inline lighthouse-core/gather/gatherers/offline.js]: I believe this technically doesn't need to be `async`, though maybe it would be best if it always was :)

File: lighthouse-core/gather/gatherers/start-url.js [hunk 2]
```diff
@@ -77,7 +93,7 @@ class StartUrl extends Gatherer {
         if (!response.fromServiceWorker) {
           return resolve({
             statusCode: -1,
-            explanation: 'Unable to fetch start URL via service worker.',
+            explanation: 'Fetched start URL but not via service worker.',
```
Comment (brendankenny) [inline lighthouse-core/gather/gatherers/start-url.js]: it seems like this message is more confusing than the original (should this audit care about that it's fetchable like that?)

Comment (patrickhulce) [reply lighthouse-core/gather/gatherers/start-url.js]: Right now the message is indistinguishable from not being able to fetch it at all which is a problem IMO. The fact that it fetched but was falling back to disk cache or something is worth flagging. Totally up for whatever wording others feel captures this goal.

File: lighthouse-core/gather/gatherers/offline.js [hunk 3]
```diff
@@ -21,14 +21,13 @@ class Offline extends Gatherer {
...
     const navigationRecord = loadData.networkRecords.filter(record => {
       return URL.equalWithExcludedFragments(record.url, passContext.url) &&
         record.fetchedViaServiceWorker;
     }).pop(); // Take the last record that matches.
 
-    return passContext.driver.goOnline(passContext)
```
Comment (paulirish) [inline lighthouse-core/gather/gatherers/offline.js]: can you add a comment around L16 that the goOnline now happens in start-url which sequentially follows this gatherer?

Comment (paulirish) [reply lighthouse-core/gather/gatherers/offline.js]: actually it seems like the goOffline in L16 isnt needed anymore since starturl does it on its own. right

Comment (patrickhulce) [reply lighthouse-core/gather/gatherers/offline.js]: > can you add a comment around L16 that the goOnline now happens in start-url which sequentially follows this gatherer?

well the `goOnline` that cleans this up actually happens in `gather-runner` that was the problem. each `afterPass` happens while online and each gatherer has to handle its own online/offline cycle

> actually it seems like the goOffline in L16 isnt needed anymore since starturl does it on its own. right

it's still needed to setup the offline for the load of the page, we do the offline check for both the page and the start url separately

File: lighthouse-core/gather/gatherers/start-url.js (hunk 2 above)
Comment (paulirish) [reply lighthouse-core/gather/gatherers/start-url.js]: > The start URL did respond, but not via a service worker

mostly trying to align with the text of the audit.. (where we dont talk about fetching, etc)

Comment (brendankenny) [reply lighthouse-core/gather/gatherers/start-url.js]: > The fact that it fetched but was falling back to disk cache or something is worth flagging

I don't know if everyone knows the priority order here, though. If I'm looking at the `start_url` audit LH tells me

> The start URL did respond, but not via a service worker

my first question would be "well why didn't LH try fetching via the SW? Do I need to force it somehow?"

It seems like the only important information is that LH tried to fetch it and the SW didn't return it. Just because it was in the disk cache during the LH run doesn't mean it will be for returning users, and isn't that the case we're testing for?

Alternatively, should we just call out the `fromDiskCache` case with a separate explanation?

File: lighthouse-core/gather/gatherers/start-url.js [hunk 4]
```diff
@@ -11,20 +11,37 @@ const Gatherer = require('./gatherer.js');
...
 class StartUrl extends Gatherer {
   /**
-   * Grab the manifest, extract it's start_url, attempt to `fetch()` it while offline
+   * Go offline, assess the start url, go back online.
    * @param {LH.Gatherer.PassContext} passContext
    * @return {Promise<LH.Artifacts['StartUrl']>}
    */
   async afterPass(passContext) {
+    // Offline pass is not actually offline in `afterPass`,
+    // each gatherer must set and cleanup its own offline state.
```
Comment (brendankenny) [inline lighthouse-core/gather/gatherers/start-url.js]: referencing the offline pass in the gatherers themselves makes me sad, though it's pretty necessary in `offline.js` at this point. For this one, though, would something like
``// `afterPass` is always online, so manually go offline to check start_url.`` or something like that work instead?

File: lighthouse-core/gather/gatherers/start-url.js [hunk 5]
```diff
@@ -11,20 +11,37 @@ const Gatherer = require('./gatherer.js');
...
+    await passContext.driver.goOffline();
+    const result = await this._determineStartUrlAvailability(passContext);
+    await passContext.driver.goOnline(passContext);
+
+    return result;
+  }
+
+  /**
+   * Grab the manifest, extract it's start_url, attempt to `fetch()` it while offline
```
Comment (brendankenny) [inline lighthouse-core/gather/gatherers/start-url.js]:
```suggestion
   * Grab the manifest, extract its start_url, attempt to `fetch()` it while offline.
```
(not your fault :)

File: lighthouse-core/gather/gatherers/start-url.js [hunk 6]
```diff
@@ -11,20 +11,37 @@ const Gatherer = require('./gatherer.js');
...
       return {statusCode: -1, explanation: startUrlInfo.reason};
     }
 
-    return this._attemptStartURLFetch(passContext.driver, startUrlInfo.startUrl).catch(() => {
-      return {statusCode: -1, explanation: 'Unable to fetch start URL via service worker.'};
-    });
+    try {
+      return await this._attemptStartURLFetch(passContext.driver, startUrlInfo.startUrl);
+    } catch (err) {
+      return {statusCode: -1, explanation: 'Unable to fetch start_url via service worker.'};
```
Comment (brendankenny) [inline lighthouse-core/gather/gatherers/start-url.js]: maybe we want "Error while attempting to fetch start_url via service worker" or something to differentiate the cases (and do we want the `err.message` if it exists?)

File: lighthouse-core/test/gather/gatherers/start-url-test.js [hunk 7]
```diff
@@ -7,168 +7,190 @@
...
-    jest.useFakeTimers();
-    baseArtifacts = {WebAppManifest: null};
+    gatherer = new StartUrlGatherer();
+    mockDriver = {
+      goOffline: unimplemented,
+      goOnline: unimplemented,
+      evaluateAsync: unimplemented,
+      on: unimplemented,
+      off: unimplemented,
+    };
```
Comment (brendankenny) [inline lighthouse-core/test/gather/gatherers/start-url-test.js]: moving off bespoke mocks is good, but I have to say this is a nice model for this file :)

File: lighthouse-core/test/gather/gatherers/start-url-test.js [hunk 8]
```diff
@@ -7,168 +7,190 @@
...
+      driver: mockDriver,
     };
-    const passContextWithFragment = {
-      url: 'https://ifixit-pwa.appspot.com/#/history',
-      baseArtifacts: createArtifactsWithURL('https://ifixit-pwa.appspot.com/#/history'),
-      driver: wrapSendCommand(mockDriver, 'https://ifixit-pwa.appspot.com/#/history', 200, true),
+
+    const result = await gatherer.afterPass(passContext);
+    expect(mockDriver.goOffline).toHaveBeenCalled();
+    expect(mockDriver.goOnline).toHaveBeenCalled();
```
Comment (brendankenny) [inline lighthouse-core/test/gather/gatherers/start-url-test.js]: we probably shouldn't assert these as we don't really care if it tried to go offline since we know it never attempts the fetch?

File: lighthouse-core/test/gather/gatherers/start-url-test.js [hunk 9]
```diff
@@ -7,168 +7,190 @@
...
 
-    return Promise.all([
-      startUrlGatherer.afterPass(passContext),
-      startUrlGathererWithFragment.afterPass(passContextWithFragment),
-    ]).then(([artifact, artifactWithFragment]) => {
-      assert.equal(artifact.statusCode, 200);
-      assert.equal(artifactWithFragment.statusCode, 200);
+    const result = await gatherer.afterPass(passContext);
+    expect(mockDriver.goOffline).toHaveBeenCalled();
+    expect(mockDriver.goOnline).toHaveBeenCalled();
```
Comment (brendankenny) [inline lighthouse-core/test/gather/gatherers/start-url-test.js]: what do you think about an extra level of checking for these that it was offline when `evaluateAsync` was called, not just that off and on were both called?

File: lighthouse-core/gather/gatherers/start-url.js (hunk 4 above)
Comment (patrickhulce) [reply lighthouse-core/gather/gatherers/start-url.js]: sure, done!

File: lighthouse-core/gather/gatherers/start-url.js (hunk 6 above)
Comment (patrickhulce) [reply lighthouse-core/gather/gatherers/start-url.js]: well this is differentiated at this point but sure message is more direct :)

File: lighthouse-core/test/gather/gatherers/start-url-test.js (hunk 9 above)
Comment (patrickhulce) [reply lighthouse-core/test/gather/gatherers/start-url-test.js]: I like it as a separate test but not on every single one :) done!
